plugin-server/src/main/ingestion-queues/anonymous-event-buffer-consumer.ts

```diff
@@ -0,0 +1,103 @@
+        query performance.
+
+        On failure to enqueue we will fail the batch. only on successfully
+        enqueuing an entire batch to Graphile Worker (a PostgreSQL backed async
+        worker library).
+
+        TODO:
+
+        1. distinguish between operational errors and programming errors
+        2. ensure offset only updated if message pushed to DLQ
+    */
+
+    const consumer = kafka.consumer({ groupId: 'ingester' })
+
+    status.info('🔁', 'Starting anonymous event buffer consumer')
+
+    await consumer.connect()
+    await consumer.subscribe({ topic: KAFKA_BUFFER })
+    await consumer.run({
+        eachBatch: async ({ batch, resolveOffset, heartbeat }) => {
+            status.info('🔁', 'Processing batch', { size: batch.messages.length })
+            for (const message of batch.messages) {
+                if (!message.value || !message.headers?.processEventAt || !message.headers?.eventId) {
+                    status.warn('⚠️', `Invalid message for partition ${batch.partition} offset ${message.offset}.`, {
+                        value: message.value,
+                        processEventAt: message.headers?.processEventAt,
+                        eventId: message.headers?.eventId,
+                    })
+                    producer.queueMessage({ topic: KAFKA_EVENTS_DEAD_LETTER_QUEUE, messages: [message] })
+                    resolveOffset(message.offset)
+                    continue
+                }
+
+                let eventPayload: PluginEvent
+
+                try {
+                    eventPayload = JSON.parse(message.value.toString())
+                } catch (error) {
+                    status.warn('⚠️', `Invalid message for partition ${batch.partition} offset ${message.offset}.`, {
+                        error,
+                    })
+                    producer.queueMessage({ topic: KAFKA_EVENTS_DEAD_LETTER_QUEUE, messages: [message] })
+                    resolveOffset(message.offset)
+                    continue
+                }
+
+                const job = {
+                    jobKey: message.headers.eventId.toString(), // Ensure we don't create duplicates
+                    eventPayload: eventPayload,
+                    timestamp: Number.parseInt(message.headers.processEventAt.toString()),
+                }
+
+                status.debug('⬆️', 'Enqueuing anonymous event for processing', { job })
+
+                // NOTE: `jobQueueManager.enqueue` handles retries internally.
+                // As such this can take a long time, so we need to ensure we
+                // keep the heartbeat going during this time, every 5 seconds.
+                // NOTE: it might be worth handling retry logic explicit here so
+                // get better control on which errors we retry on.
+                const heartbeatTimer = setInterval(async () => await heartbeat(), 5000)
+                await jobQueueManager.enqueue(JobName.BUFFER_JOB, job)
```

Comment:
I think there's a misunderstanding here - the manager does handle retries internally, but it will return after the first try, success or not. Thus this heartbeat timer is unnecessary and we're actually not getting a guarantee of something being enqueued at the moment. 

With this Kafka queue I'd effectively get rid of the in-memory queue and just rely entirely on Kafka for the buffering/retries. You just need to make sure the consumer doesn't crash.

Reply:
Ah ok I'll have a closer look at that code then. Ideally I want something that fails so we can let the KafkaJs retry logic kick in.

Reply:
Exactly

Reply:
I've refactored to use `GraphileWorker.enqueue` directly now.  Does this look right?

Reply:
Well... Not under the current approach but I think this might be the right direction to go in.

Essentially everything atm _should_ go through the job queue manager. However, that's something that was built in the initial iteration of job queues and never really used. The idea was, as I've mentioned before, to have tiers of queues to make things robust. i.e. memory, Postgres, S3.

However, we are departing from that approach with this Kafka queue, so might even make sense to yeet this whole management code. 

All in all, the change is fine - just being explicit here about what we're doing as that very line change signifies a departure from the previous architecture :D

Reply:
Great.

Yes not a fan of the tiered system. Seems like a big complexity explosion. That seems like a whole product, why aren't there services already providing this, seems like a red flag. :100: we should yeet that and rely on characteristics that established tools give us imo. There are communities dedicated to this, we can use these and handle our own product and not get distracted ✅